src/thunderbots/software/ai/primitive/shoot_primitive.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "ai/primitive/primitive.h"
+#include "geom/angle.h"
+#include "geom/point.h"
+
+class ShootPrimitive : public Primitive
+{
+   public:
+    static const std::string PRIMITIVE_NAME;
+    /**
+     * Creates a new Shoot Primitive
+     *
+     * @param robot_id The id of the Robot to run this Primitive
+     * @param shot_origin The location where the shot will be taken
+     * @param shot_direction The orientation the Robot will shoot at
+     * @param power The power of the kick (how fast the Robot will kick the ball)
+     * or chip (distance between start location of chipping and first bounce)
+     * @param chip Whether to chip the ball or not (If true, the Robot will chip,
+     * if false, the Robot will kick)
+     */
+    explicit ShootPrimitive(unsigned int robot_id, const Point &shot_origin,
+                            const Angle &shot_direction, double power, bool chip);
+
+    /**
+     * Creates a new Shoot Primitive from a Primitive message
+     *
+     * @param primitive_msg The message from which to create the Shoot Primitive
+     */
+    explicit ShootPrimitive(const thunderbots_msgs::Primitive &primtiive_msg);
+
+    std::string getPrimitiveName() const override;
+
+    unsigned int getRobotId() const override;
+
+    std::vector<double> getParameterArray() const override;
+
+    std::vector<bool> getExtraBitArray() const override;
+
+   private:
+    unsigned int robot_id;
+    Point shot_origin;
+    Angle shot_direction;
+    double power;
```

Comment:
We had some discussion about using `power` on a recent PR: https://github.com/UBC-Thunderbots/Software/pull/108#discussion_r226637254
(Tl;DR - power can be a bit of a confusing word (ie. it's technically work per unit time))

I'm just guessing at what this is supposed to be, but maybe `kick_speed_meters_per_second` would be a bit more descriptive?

Reply:
Yea, I'd hesitate a bit more here to call it `kick_speed_meters_per_second` because the units are not always meters per second (aka speed). I was ok with `kick_speed_meters_per_second` in the context of grsim because `m/s` are the units for both kicking and chipping.

For our robots, the units when referring to a kick are `m/s` for the speed of the kick. When we are chipping, the units are just `meters`, for the distance between the robot and the location of the first bounce of the ball.

The different units in different cases make it harder to name, and I'd also rather not split this into 2 variables (eg. `kick_speed_meters_per_second` and `chip_distance_meters`) because you'd effectively have a duplicate variable which also adds complexity.

I summon @EvanMorcom @henrymliu @akhilveeraghanta and @jordanzliu for their opinions! :crystal_ball:

Reply:
Hmm, I agree about not splitting variables, but looking at this from somewhat of an "external" prespective, it feels like we should have two primitives for `chip`, and `kick` respectively. For that matter, what differentiates `shoot` from just a kick?

Reply:
I agree that `power` can be more descriptive since there are two different possible functionalities to it. Is there any possible way to, say, create the variable under a conditional statement (is that how they call `if` and stuff?) after determining whether it's a `chip` or a `kick`? Or would that be even more complicated than setting up two separate variables.

Reply:
Thanks all for the feedback, by the way!

Reply:
This seems like a textbook example of logical cohesion. We shouldn't have a flag that determines what the `power` variable means imo, we should definitely split `shoot` into `kick` and `chip` primitives as Gareth suggested, and give them their own definitions of `power`.

Reply:
I think I'll tentatively agree with splitting the primitives. Shouldn't be too much work on the high level side.

Reply:
>     1. Our current radio protocol only support 8 unique primitives. We should make sure we can still support all the other primitives we need if we are adding another one. This could change if/when we move to wifi or a better radio, but that won't be for at least a year.

I don't think limiting ourselves in terms of what we can do in `ai` based on the limitations of the radio protocol like this isn't a great idea. I'd be ok with encoding them into the "same" primitive (though it's still not a great solution). No chance of just changing the radio protocol I suppose?

>     2\. A bit more nitpicky, but we'll want a better way to share code on the robot firmware. Seems like the implementations for `shoot` and `chip` on the robot would be exactly the same except for the chip/kick flag

This doesn't seem like it's really related to this PR? More of a firmware implementation thing?

Reply:
> This seems like a textbook example of logical cohesion. We shouldn't have a flag that determines what the `power` variable means imo, we should definitely split `shoot` into `kick` and `chip` primitives as Gareth suggested, and give them their own definitions of `power`.

I Agree with this.

Reply:
I have tentatively split `shoot` into separate `kick` and `chip` primitives (didn't remove the files for `shoot`, just in case). It's true that the implementations are basically exactly the same except for the content of `power`. 
Just out of curiosity, what are the downsides of not having a one-to-one mapping between a high-level primitive and a firmware primitive?

Reply:
I might be wrong here (@henrymliu @MathewMacDougall please correct me if I am), but I think it just makes it a bit easier for people to reason about things. For example, if I send a high-level `kick` primitive to the robot and it's not working correctly, the first place I might think to go look is in the `kick` primitive on the robot side. When there's not a one-to-one mapping however, I might need to trace the path that the `kick` primitive is taking, see that it gets mapped to a `shoot` primitive, and _then_ could figure out where in the robot firmware I could look for the potential error.

Reply:
What @garethellis0 said is correct. When things have a 1 to 1 mapping it's simpler and easier to reason about the flow of control and what changes affect what. I consider mapping two high-level primitives to a single firmware primitive a "trick", and the more tricks code has the harder it is to fully understand (quickly). However I do think this is a good solution.

Reply:
^ I'm also super happy to see this kind of discussion happening with so many people :grin:. It's great to see everyone working together to learn and get to the best solution.